### PR TITLE
feat: nip-42

### PR DESCRIPTION
**Pull request by npub1zenn0**
`relay.ts` had some type errors, it wasn't even building for me out of the box so I went and included a fix in this MR.

Fixes https://github.com/nbd-wtf/nostr-tools/issues/189

### CODE REVIEW

File: nip42.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  const e: EventTemplate = {
+    kind: Kind.ClientAuth,
+    created_at: Math.floor(Date.now() / 1000),
+    tags: [
+      ['relay', relay.url],
+      ['challenge', challenge]
+    ],
+    content: ''
+  }
+  const sub = relay.publish(await sign(e), 'AUTH')
```
Comment (fiatjaf) [inline nip42.ts:31]: this is not a subscription

File: relay.ts [hunk 2]
```diff
@@ -334,24 +348,27 @@ export function relayInit(
...
         }, countTimeout)
-        s.on('count', (event: Event) => {
+        s.on('count', (event: CountPayload) => {
           s.unsub()
           clearTimeout(timeout)
           resolve(event)
         })
       }),
-    publish(event: Event): Pub {
+    publish(event, type = 'EVENT'): Pub {
```
Comment (fiatjaf) [inline relay.ts:367]: can you make a separate function `auth()` instead? I think it is clearer.

**Comment by fiatjaf** [comment]
Thank you, this looks good, and thanks for fixing other stuff too.

**Comment by npub1zenn0** [comment]
Done

File: nip42.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    created_at: Math.floor(Date.now() / 1000),
+    tags: [
+      ['relay', relay.url],
+      ['challenge', challenge]
+    ],
+    content: ''
+  }
+  const pub = relay.auth(await sign(e))
+  return new Promise((resolve, reject) => {
+    pub.on('ok', function ok() {
```
Comment (npub1zenn0) [inline nip42.ts:33]: oh yea should I have called it "auth" now? :grin: didn't occur to me

**Pull request merged**